Dockerfile

```diff
@@ -0,0 +1,30 @@
+FROM node:14.15-alpine as build
+
+## Variables
+# User and ID
+ARG APP=brasil-api 
+ARG ID=1001
+
+# Default application port "from next framework"
+ARG PORT=3000
+
+
+## Commands
+# Add User
+RUN     adduser --uid ${ID} -D -h /home/${APP} ${APP} node 
+# Set User
+USER ${APP}
+#Copy files
+COPY --chown=${ID}:node . /home/${APP}
+# Set work directory
+WORKDIR /home/${APP}
+# Install dependencies, build project
+RUN     npm -l install npm \
+    &&  npm update \
+    &&  npm install \
+    &&  npm run build
+# Expose default port
+EXPOSE ${PORT}
+
+# The command to be executed when this conteiner be deploy
+ENTRYPOINT [ "npm", "run", "start" ]
```

Comment:
Então, aqui deveria ser `CMD [ "npm", "run", "start" ]` mesmo. Não vejo necessidade de usar `ENTRYPOINT` nesse caso.

Um ponto que deve ser levado em conta quando for criar uma imagem para produção é pensar que a imagem vai ser para produção. Colocar opções para criar usuário, instalar ferramentas de desenvolvimento, acesso shell, etc. Tudo isso é desnecessário. Tente fazer a imagem mais simples possível para que ela não vire um megazord.

E quanto a imagens multistage, o último estágio deve ser o mais clean possível. Como já foi feito a instalação de dependências, build, etc. em outros estágios, só copie o necessário, instale alguma ferramenta (next) que vai rodar o servidor e rode o servidor.

Reply:
Sim, Acho que esse comentário fala de dois ficheiros diferentes.

Um para uso exclusivo do vscode ou afins. *Dockerfile.vscode* é um para runtime.
Para montar um ambiente genérioco de desenvolvimento.

O Dockerfile, esse sim é para produtivo.

No caso do temos a hierarquia de uso .:
> RUN  -> CMD  ->  ENTRYPOINT

No caso do *CMD* ele é utilizado para executar algum comando em runtime, ex .:
- correr ficheiros de configuração
- ler variaveis de ambiente
- iniciar serviços auxiliares

No geral preparar o ambiente para que a minha aplicação possa correr da melhor forma.

Depois o *ENTRYPOINT*, ele é a ultima coisa que o meu conteiner fará em runtime, e que é usado como healthcheck. Caso o *ENTRYPOINT* pare o conteiner reinicia ou emite erro, um *CMD* não.
- iniciar a base de dados
- iniciar o servidor
- correr um serviço
- ...

E isso ocupa o terminal.
Assim como subir o servidor é o principal o *ENTRYPOINT* é o recomendado.